registry/repository.go

```diff
@@ -187,13 +203,38 @@ func (c *repositoryClient) getSignatureBlobDesc(ctx context.Context, sigManifest
 
 // uploadSignatureManifest uploads the signature manifest to the registry
 func (c *repositoryClient) uploadSignatureManifest(ctx context.Context, subject, blobDesc ocispec.Descriptor, annotations map[string]string) (ocispec.Descriptor, error) {
-	opts := oras.PackOptions{
+	configDesc, err := pushNotationManifestConfig(ctx, c.GraphTarget)
+	if err != nil {
+		return ocispec.Descriptor{}, fmt.Errorf("failed to push notation manifest config: %w", err)
+	}
+
+	opts := oras.PackManifestOptions{
 		Subject:             &subject,
 		ManifestAnnotations: annotations,
-		PackImageManifest:   true,
+		Layers:              []ocispec.Descriptor{blobDesc},
+		ConfigDescriptor:    &configDesc,
 	}
 
-	return oras.Pack(ctx, c.GraphTarget, ArtifactTypeNotation, []ocispec.Descriptor{blobDesc}, opts)
+	return oras.PackManifest(ctx, c.GraphTarget, oras.PackManifestVersion1_1_RC4, "", opts)
+}
+
+// pushNotationManifestConfig pushes an empty notation manifest config, if it
+// doesn't exist.
+func pushNotationManifestConfig(ctx context.Context, pusher content.Storage) (ocispec.Descriptor, error) {
+	// check if the config exists
+	exists, err := pusher.Exists(ctx, notationEmptyConfigDesc)
+	if err != nil {
+		return ocispec.Descriptor{}, fmt.Errorf("unable to verify existence: %s: %s. Details: %w", notationEmptyConfigDesc.Digest.String(), notationEmptyConfigDesc.MediaType, err)
+	}
+	if exists {
+		return notationEmptyConfigDesc, nil
+	}
+
+	// push the config
+	if err := pusher.Push(ctx, notationEmptyConfigDesc, bytes.NewReader(notationEmptyConfigData)); err != nil && !errors.Is(err, errdef.ErrAlreadyExists) {
+		return ocispec.Descriptor{}, fmt.Errorf("unable to push: %s: %s. Details: %w", notationEmptyConfigDesc.Digest.String(), notationEmptyConfigDesc.MediaType, err)
+	}
```

Comment:
What if `err` is `errdef.ErrAlreadyExists`? We are ignoring this error.
Since we already checked existence above, can I say `!errors.Is(err, errdef.ErrAlreadyExists)` is always true? If yes, then, how about,
```suggestion
	if err := pusher.Push(ctx, notationEmptyConfigDesc, bytes.NewReader(notationEmptyConfigData)); err != nil {
		return ocispec.Descriptor{}, fmt.Errorf("unable to push: %s: %s. Details: %w", notationEmptyConfigDesc.Digest.String(), notationEmptyConfigDesc.MediaType, err)
	}
```

Reply:
We do want to ignore the `errdef.ErrAlreadyExists` error as it is not helpful for the user, it's fine as long as the config is there. It is also possible that `Exists` returns `false` at the time we checks it, but another process uploads a same config before we do `Push`.

Reply:
Then we should leave a comment in the code stating this. Currently, the function could return `nil` even though the error is `non-nil`.

Reply:
Added comments.